src/telemetry_poller.erl

```diff
@@ -0,0 +1,407 @@
+%% '''
+%%
+%% <blockquote>Note: the other solution would be to dispatch two different events by hooking up
+%% `example_app:regular_users_session_count/0' and `example_app.admin_users_session_count/0'
+%% functions directly. However, if you add more and more user roles to your app, you'll find
+%% yourself creating a new event for each one of them, which will force you to modify existing
+%% event handlers. If you can break down event value by some feature, like user role in this
+%% example, it's usually better to use event metadata than add new events.
+%% </blockquote>
+%%
+%% This is a perfect use case for poller, because you don't need to write a dedicated process
+%% which would call these functions periodically. Additionally, if you find that you need to collect
+%% more statistics like this in the future, you can easily hook them up to the same poller process
+%% and avoid creating lots of processes which would stay idle most of the time.
+%% @end
+%%%---------------------------------------------------
+-module(telemetry_poller).
+
+-behaviour(gen_server).
+
+%% API
+-export([
+    child_spec/1,
+    list_measurements/1,
+    start_link/1,
+    stop/1,
+    stop/2,
+    stop/3
+]).
+
+-export([code_change/3, handle_call/3, handle_cast/2,
+     handle_info/2, init/1, terminate/2]).
+
+-define(DEFAULT_VM_MEASUREMENTS, [memory, total_run_queue_lengths]).
+
+-ifdef('OTP_RELEASE').
+-include_lib("kernel/include/logger.hrl").
+-else.
+-define(LOG_ERROR(Msg, Args), error_logger:error_msg(Msg, Args)).
+-endif.
+
+-ifdef('OTP_RELEASE').
+-include_lib("kernel/include/logger.hrl").
+-else.
+-define(LOG_WARNING(Msg, Args), error_logger:warning_msg(Msg, Args)).
+-endif.
+
+-ifdef('OTP_RELEASE').
+-define(WITH_STACKTRACE(T, R, S), T:R:S ->).
+-else.
+-define(WITH_STACKTRACE(T, R, S), T:R -> S = erlang:get_stacktrace(),).
+-endif.
+
+-type t() :: gen_server:server().
+-type options() :: [option()].
+-type option() ::
+    {name, gen_server:name()
+    | {period, period()}
+    | {measurements, [measurement()]}
+    | {vm_measurements, default | [vm_measurement()]}}.
+-type measurement() :: {module(), atom(), []}.
+-type period() :: pos_integer().
+-type vm_measurement() ::
+    memory
+    | total_run_queue_lengths.
+-type state() :: #{measurements => [measurement()], period => period()}.
+
+%% @doc Starts a poller linked to the calling process.
+%%
+%% Useful for starting Pollers as a part of a supervision tree.
+-spec start_link(options()) -> gen_server:on_start().
+start_link(Opts) when is_list(Opts) ->
+    Name = proplists:get_value(name, Opts, ?MODULE),
+    Args = parse_args(Opts),
+    gen_server:start_link({local, Name}, ?MODULE, Args, []).
+
+%% @doc
+%% Stops the poller with specified reason.
+%%
+%% See documentation for {@link gen_server:stop/3} to learn more about the behaviour of this function.
+-spec stop(t()) -> ok.
+stop(Poller) ->
+    stop(Poller, normal, inifinity).
```

Comment:
Typo on `infinity` here and below. Is this function being tested? :)

Reply:
Actually, let's just kill the `stop/1/2/3` functions. We can always expose it later if someone needs it. :)

Reply:
I'm all in favor of deleting unnecessary code :)